hawkular-command-gateway-war/src/main/java/org/hawkular/cmdgw/ws/server/FeedListenerGenerator.java

```diff
@@ -1,165 +0,0 @@
- * See the License for the specific language governing permissions and
- * limitations under the License.
- */
-package org.hawkular.cmdgw.ws.server;
-
-import java.util.ArrayList;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
-import javax.annotation.PostConstruct;
-import javax.annotation.Resource;
-import javax.ejb.EJB;
-import javax.ejb.Singleton;
-import javax.ejb.Startup;
-import javax.ejb.TransactionAttribute;
-import javax.ejb.TransactionAttributeType;
-import javax.enterprise.concurrent.ManagedExecutorService;
-import javax.jms.ConnectionFactory;
-import javax.naming.InitialContext;
-import javax.persistence.PostRemove;
-
-import org.hawkular.bus.common.ConnectionContextFactory;
-import org.hawkular.bus.common.Endpoint;
-import org.hawkular.bus.common.MessageProcessor;
-import org.hawkular.bus.common.consumer.ConsumerConnectionContext;
-import org.hawkular.cmdgw.ws.Constants;
-import org.hawkular.cmdgw.ws.MsgLogger;
-import org.hawkular.cmdgw.ws.mdb.AddDatasourceListener;
-import org.hawkular.cmdgw.ws.mdb.AddJdbcDriverListener;
-import org.hawkular.cmdgw.ws.mdb.DeployApplicationListener;
-import org.hawkular.cmdgw.ws.mdb.ExecuteOperationListener;
-import org.hawkular.cmdgw.ws.mdb.ExportJdrListener;
-
-@Startup
-@Singleton
-@TransactionAttribute(TransactionAttributeType.NOT_SUPPORTED)
-public class FeedListenerGenerator {
-    @EJB
-    private ConnectedFeeds connectedFeeds;
-
-    @Resource(mappedName = Constants.CONNECTION_FACTORY_JNDI)
-    private ConnectionFactory connectionFactory;
-
-    private Map<String, ConnectionContextFactory> connContextFactories;
-    private Map<String, List<ConsumerConnectionContext>> consumerContexts;
-
-    @Resource
-    private ManagedExecutorService threadPoolService;
-
-    @PostConstruct
-    public void initialize() throws Exception {
-        if (this.connectionFactory == null) {
-            MsgLogger.LOG.warnf("Injection of ConnectionFactory is not working - looking it up explicitly");
-            InitialContext ctx = new InitialContext();
-            this.connectionFactory = (ConnectionFactory) ctx.lookup(Constants.CONNECTION_FACTORY_JNDI);
-        } else {
-            MsgLogger.LOG.warnf("Injection of ConnectionFactory works - you can remove the hack");
-        }
-
-        connContextFactories = new HashMap<>();
-        consumerContexts = new HashMap<>();
-    }
-
-    @PostRemove
-    public void shutdown() throws Exception {
-        if (connContextFactories != null) {
-            for (String feedId : this.connContextFactories.keySet()) {
-                removeListeners(feedId);
-            }
-        }
-    }
-
-    /**
-     * @return the connection factory this object will use when connecting to the messaging system.
-     */
-    public ConnectionFactory getConnectionFactory() {
-        return this.connectionFactory;
-    }
-
-    public void addListeners(String feedId) throws Exception {
```

Comment:
Ok, this is a major change removing the way the listeners were added, and I believe the design is now broken. This was an important feature of the listener generator that has been removed, but I think it needs to be put back.

Now, it appears the new code only adds the bus listeners at application startup. However, this breaks the messaging when we run hawkular clustered (i.e. more than one hawkular server participating in the bus cluster).

Here's the use-case. We have N hawkular servers clustered (where N > 1). The way the code is now (if I am reading it right, correct me if I'm wrong) as soon as each hawkular server starts up, they will bind listeners to the bus queues (e.g. so there are N listeners on queue "FeedCommandQueue" for example). So we have N listeners even though NO feeds are currently connected! So now suppose a UI sends a request to deploy an app to a feed. ALL the N servers will now have a chance to pick that message off the queue - even though, again, NO feed is connected! This means one of the N servers will get the message, and spit out an error saying "there is no feed" and the UI gets an error. Doesn't sound too bad until...

Now suppose a feed connects - say it connects to server 1 (suppose we have 3 servers clustered). Even though the feed is connected to server #1, servers #2 and #3 have a chance to pick up the message since they have listeners created at startup. The UI sends a message. Any server's listener will have a change to pick up the queue message - say, server #2 gets it. server #2 isn't connected to the feed! Server #1 is. So the UI gets an error even though the feed is up and ready to accept commands.

We need to dynamically add and remove listeners to the server that the feed connects/disconnects from. So, at startup, NO servers have listeners to feeds. If a UI sends a command, that message is queued and will await the feed to connect. So suppose a feed connects to server #1. Server #1 will immediately add a listener to the queue (listening with message selector that only selects messages for that feed). The listener will immediately get the message queued earlier, send the message to the feed, and the command workflow continues. If the feed discnnects, the listener is removed, and again, no server will have a listener that can process the message.

So, in short, we need to have the listeners dynamically added and removed as feeds are connected/disconnected. We cannot have all the servers connecting the listeners at startup.